### ISSUE TITLE
Mark Variable as `.internal()`

### ISSUE THREAD

**Issue by coltmcnealy-lh** [littlehorse-enterprises/littlehorse#582]
### Context

I recently implemented a major overhaul to `WfSpec` versioning in which we added the `majorVersion` and `revision` fields. A new `majorVersion` happened when there was a breaking change, defined as a change in the set of `searchable()` or `required()` variables.

### Description

What if there is a `Variable` that is not searchable (for performance reasons), and not an input variable, but it _is_ an important part of the Contract of the `WfSpec` (for example when using `WfRun`'s to store business entitites, this variable might be a useful "column" but not indexed or required)?

I think we should make any changes to the sets of `Variable`s in a `WfRun` result in a new `majorVersion`, unless the variable is marked as `internal()`.

All three SDK's should have a `WfRunVariable#internal()` method

### Technical Context

Certain variables are automatically created by the SDK, and should also be marked as `internal()`. For example:

- internal variables created by the `WorkflowThread#spawnThread()` method
- The `INPUT` variables used for receiving `ExternalEvent`'s in an Interrupt Handler, and `Failure`'s in a Failure Handler

**Comment by coltmcnealy-lh** [comment]
@eduwercamacaro agrees; also we should should update our documentation. This should be done before `1.0.0` as it does break our contract a little bit (but not a huge amount)

**Comment by coltmcnealy-lh** [comment]
Well, it looks like this requires some discussion to get the correct API semantics, but no one wants to do the implementation 😂

**Comment by coltmcnealy-lh** [comment]
Okay, I looked at the protobuf structure. #605 introduced the `enum WfRunVariableAccessLevel` as follows:

```
// Determines the Access Level for a Variable in a ThreadSpec/WfSpec.
enum WfRunVariableAccessLevel {
  // A `PUBLIC_VAR` can be accessed (read + mutated) by child `WfRun`'s.
  PUBLIC_VAR = 0;

  // A `PRIVATE_VAR` cannot be accessed by a child `WfRun`.
  PRIVATE_VAR = 1;

  // An `INHERITED_VAR` is inherited from the parent `WfRun`. Only valid in a `WfSpec` that
  // has a parent. Also can only be declared in the Entrypoint Thread.
  INHERITED_VAR = 2;
}
```

I was originally going to add a `bool internal = 6;` to the `message ThreadVarDef`. However, that structure matches perfectly with the motivation of this ticket. We need to do the following:

- [ ] Make variables automatically declared by the SDK use the `PRIVATE_VAR` level (eg. variables declared for storing the output of a `START_THREAD` node).
- [ ] Ignore `PRIVATE_VAR` variables when determining `WfSpec` version compatibility.

**Comment by coltmcnealy-lh** [comment]
It appears we need a little more definition of how the `WfRunVariableAccessLevel` and the `frozenVarDefs` play together. The `WfRunVariableAccessLevel` has three possible values:
* `PUBLIC_VAR` -> the variable is part of the public API of the `WfSpec`, and if it changes then the major version is incremented. Child workflows can inherit it.
* `PRIVATE_VAR` -> The variable is not part of the public API of the `WfSpec` and clients should not rely on it as part of a data model. Child workflows cannot inherit it.
* `INHERITED_VAR` -> The variable comes from a parent workflow.

We also have the concept of `frozen_var_defs`. These variables are "frozen" into a specific type, meaning that future versions of the same `WfSpec` can never have a `VariableDef` with the same name and a different type. This is nice for a few reasons:
- `rpc SearchVariable` allows searching by `wfSpecName` and `var_name` without specifying a `wf_spec_version`. This is useful for people using LH as a data store. This means that  
- We plan (#584) to allow users to configure a stream of output events into a Kafka topic on a per-tenant basis. One of the options will be to treat a `WfSpec` as a column, and a `WfRun` as a row, in a database (an example use-case of these messages would be materializing it to an Iceberg table such as Confluent Tableflow does). Dealing with different types for the same variable makes Confluent's implementation of TableFlow much more complicated. Frozen vardefs removes that confusion.

### Basic Options

1. Remove the `frozen_var_defs` field completely.
2. Make all `PUBLIC` variables `frozen`. 
3. Keep the same strategy (only `searchable` variables become `frozen`), and exclude `PRIVATE_VAR`s from the `frozen` set.

I am partial to option `2` because it would make the `rpc SearchVariable` and the Output Topic / Workflow Streams user experience the best.

However, option `3` is best for developer experience for _basic_ features, since only advanced use-cases will rely on `rpc SearchVariable` in a non-debugging scenario, and new users might be confused as to why they cannot change the type of a variable. (Additionally, the default access level for a variable is `PUBLIC_VAR`, not `PRIVATE_VAR`, so changing it 

### Advanced Option

After writing the above, I realized that the best option is the following:
* Make the default access level `PRIVATE_VAR`.
* Only look at `PUBLIC_VAR`s when determining whether a new `WfSpec` is a `majorVersion` or `minorRevision`.
* All `PUBLIC_VAR`s become frozen (i.e. you can never change their type).

Note that we are safe from any confusing results for basic users because:
- No variables will become frozen unless they use the _opt-in_ `.public()` feature.
- By default, child workflows won't be able to mutate a variable from the parent, which is a good safeguard.
- Using LittleHorse as a data store is an advanced use-case which most of our users won't use anyways; therefore, putting it behind an opt-in feature (using `.public()` / `.withAccessLevel(PUBLIC_VAR)`) is fine. And the implementation of `.public()` as described above gives the best possible user-experience for this advanced use-case.

**Comment by mijailr** [comment]
I believe `.internal()` should be the default behavior of every variable and we should be more specific about the variables that can be set from the entrypoint so is clearer which variables are either required or optional while running a wfRun.

**Comment by coltmcnealy-lh** [comment]
Fixed by #1118
